client.go

```diff
@@ -74,6 +156,141 @@ func SelectOneOf(protos []string, rwc io.ReadWriteCloser) (string, error) {
 	return "", ErrNotSupported
 }
 
+func simOpen(protos []string, rwc io.ReadWriteCloser) (string, bool, error) {
+	mynonce := make([]byte, 32)
+	_, err := rand.Read(mynonce)
+	if err != nil {
+		return "", false, err
+	}
+
+	werrCh := make(chan error, 1)
+	go func() {
+		myselect := []byte(tieBreakerPrefix + base64.StdEncoding.EncodeToString(mynonce))
+		err := delimWriteBuffered(rwc, myselect)
+		werrCh <- err
+	}()
+
+	// skip exactly one protocol
+	// see https://github.com/multiformats/go-multistream/pull/42#discussion_r558757135
+	_, err = ReadNextToken(rwc)
+	if err != nil {
+		return "", false, err
+	}
+
+	// read the tie breaker nonce
+	tok, err := ReadNextToken(rwc)
+	if err != nil {
+		return "", false, err
+	}
+	if !strings.HasPrefix(tok, tieBreakerPrefix) {
+		return "", false, errors.New("tie breaker nonce not sent with the correct prefix")
+	}
+
+	if err = <-werrCh; err != nil {
+		return "", false, err
+	}
+
+	peernonce, err := base64.StdEncoding.DecodeString(tok[7:])
+	if err != nil {
+		return "", false, err
+	}
+
+	var iamserver bool
+	switch bytes.Compare(mynonce, peernonce) {
```

Comment:
Will this yield the correct result if one peer sends a "2" and the other one a "0000001"?

On a more general note (and as I've argued on the specs PR), I still believe that using a 256-bit number here is excessive. We definitely can live with a collision probability of 2^-64 (for that matter, 2^-32 would be totally fine as well, and might simplify stuff in JS, where 64 bit numbers tend to be more problematic), which would allow us to use a simple `uint64` / `uint32` comparison here.

Reply:
why? It doesn't cost anything to use bigger nonces, which makes the probability of collision astronomically low.

Reply:
Why? Because it simplifies the code. I'm still not sure if `bytes.Compare` will give the correct result if the number is encoded with leading 0s.

With a 32 bit nonce the chance is already astronomically small. At this probability, you'd expect a collision every 2^31 connection attempts. Assuming you're performing 10 simultaneous connects **per second** (I don't think any node will ever do this), you'll have to wait for almost 7 years until you get the first collision. With a 64 bit nonce, you'd have to wait about 30 billion years, twice the age of the universe.

Reply:
@vyzo @marten-seemann 

I think the idea of using a random uint64 is a reasonable one and have made the change in the PR. Please let me know if there are any **strong** objections against this and we can discuss this in depth.

Reply:
@marten-seemann the code was correct. It took a random byte string, encoded it, decoded it, then compared the bytes.

Honestly, that's a _lot_ simpler than comparing a 64bit number as we're not relying on encoding order at all.

However, sending a raw number also works and I agree that 32 bytes was overkill for our purposes.